kong/pdk/node.lua

```diff
@@ -255,6 +257,23 @@ local function new(self)
     return gsub(hostname, "\n$", "")
   end
 
+
+  local prefix = self and self.configuration and self.configuration.prefix
```

Comment:
I can't help but feel like _all_ of the logic should go into `kong.pdk.private.node` instead. As it is, it's still hard to follow the order of operations to know if there's a chicken/egg situation.

Here's how I picture the call stack:

```
kong.init()
 -> kong.global.init_pdk()
  -> kong.pdk.new()
    -> kong.pdk.node.new()
      -> kong.pdk.private.node.init()
```

...and `kong.pdk.node.private.init()` could look something like this:

```lua
-- init() is a no-op if called more than once
if already_initialized() then
  return
end

-- check shm first, since this might have been a reload
local shm_id = shm:get(MY_ID_KEY)

local fs_id = read_id_from_filesystem()

if shm_id and not fs_id then
  store_to_filesystem(shm_id)

elseif fs_id and not shm_id then
  store_to_shm(fs_id)

elseif not fs_id and not shm_id then
  local id = generate_uuid()
  store_to_filesystem(id)
  store_to_shm(id)

elseif fs_id ~= shm_id then
  -- uh oh, need to pick a winner
  -- probably should keep shm_id in this case because if it exists,
  -- that means we're a reload, so it's already being used by the
  -- current process group
  log("uh oh this is weird...")
  store_to_filesystem(shm_id)

else
  -- no action needed if this is the case
  assert(fs_id == shm_id)
end
```

At this point the `kong.node.get_id()` PDK function just needs to be updated to not attempt to create the node_id itself:

```diff
  function _NODE.get_id()
    if node_id then
      return node_id
    end

    local shm = ngx.shared.kong

-    local ok, err = shm:safe_add(NODE_ID_KEY, utils.uuid())
-    if not ok and err ~= "exists" then
-      error("failed to set 'node_id' in shm: " .. err)
-    end

    node_id, err = shm:get(NODE_ID_KEY)
    if err then
      error("failed to get 'node_id' in shm: " .. err)
    end

    if not node_id then
      error("no 'node_id' set in shm")
    end

    return node_id
  end
```

Reply:
@flrgh This is a cleaner approach. But there may have an issue:

```
kong.init()
 -> kong.global.init_pdk(self, kong_config)
  -> kong.pdk.new(kong_config, self)
    -> kong.pdk.node.new(self)
      -> kong.pdk.private.node.init(self)
```

Kong calls `kong.global.init_pdk` in the cmd stage, and the `store_to_shm` function won't work as `ngx.shared` API is not available in the stage.